File: tests/python/test_map_attribute.py
# SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
# SPDX-License-Identifier: LicenseRef-NvidiaProprietary
#
# Public ovstage map/unmap test: map_attribute reserves writable storage-side
# buffers you fill directly (the zero-copy programming model), committed per group
# via unmap_group or all at once by the final unmap. Covers an existing column and
# a freshly-created one. CPU-only. The write-flavors example is the workflow tour.

import numpy as np
import pytest

from ovstage import DLDataType, DLDataTypeCode, OrdinalRange, PathDictionary

FLOAT3 = DLDataType(code=DLDataTypeCode.kDLFloat, bits=32, lanes=3)

# Only numpy >= 2.1 requests a versioned DLPack capsule and honors its read-only
# flag. Older numpy takes the legacy capsule, which has no flags field at all, so a
# readonly=True export is silently writable there and the rejection cannot be tested.
_NUMPY_HONORS_READONLY = np.lib.NumpyVersion(np.__version__) >= "2.1.0"


def _rows_by_prim(stage, query, attr, end_ordinal, lanes):
    """Latest committed value(s) per prim (list-relative index -> list of floats)."""
    read = stage.read_attributes(query, [attr], OrdinalRange.latest(end_ordinal))
    read.wait()
    rows = {}
    for group in read.groups():
        if not group.is_delete:
            values = np.asarray(group.array(0)).reshape(-1)
            for local in range(group.prim_count):
                row = group.data_row_index(local) if group.has_data_index_map else local
                rows[group.prim_index(local)] = [float(v) for v in values[row * lanes:row * lanes + lanes]]
        stage.release_group(group)
    read.release().wait()
    return rows


def test_map_existing_and_fresh_column(stage):
    with PathDictionary(stage) as paths:
        order = ["/World/Mapped/A", "/World/Mapped/B"]
        plist = paths.create_path_list_from_strings(order)
        query = stage.query_from_path_list(plist)
        try:
            existing = paths.intern_token("map-existing")
            fresh = paths.intern_token("map-fresh")
            stage.write_attribute(
                query, existing, ordinal=1, tensors=np.array([1.0, 2.0], np.float32), is_array=False
            ).wait()
            stage.advance_write_floor(ordinal=1).wait()

            # [snippet:map-unmap-cpu]
            # map_attribute reserves storage-side buffers you fill directly. The
            # mapped buffer is a WRITE-ONLY staging buffer for the session's ordinal
            # (uninitialized, not a view of current values), so fill every element.
            # Commit per group via unmap_group (streaming) or all at once when the
            # session's with-block exits (the final unmap).
            fill = {0: 10.0, 1: 20.0}  # by list-relative prim index (A, B)
            with stage.map_attribute(query, existing, ordinal=2) as session:
                session.wait()
                for group in session.groups():
                    buffer = group.array(0)
                    for local in range(group.prim_count):
                        row = group.data_row_index(local) if group.has_data_index_map else local
                        buffer[row] = fill[group.prim_index(local)]
                    session.unmap_group(group).wait()  # streaming commit of this group
            stage.advance_write_floor(ordinal=2).wait()

            # A NEW column: the descriptor dtype (3-lane float32) defines it. No
            # unmap_group this time — the final unmap commits everything.
            fill3 = {0: [1.0, 2.0, 3.0], 1: [4.0, 5.0, 6.0]}
            with stage.map_attribute(query, fresh, ordinal=3, dtype=FLOAT3) as session:
                session.wait()
                for group in session.groups():
                    buffer = group.array(0)
                    for local in range(group.prim_count):
                        row = group.data_row_index(local) if group.has_data_index_map else local
                        buffer[row * 3:row * 3 + 3] = fill3[group.prim_index(local)]
            stage.advance_write_floor(ordinal=3).wait()
            # [/snippet:map-unmap-cpu]

            assert _rows_by_prim(stage, query, existing, 2, 1) == {0: [10.0], 1: [20.0]}
            assert _rows_by_prim(stage, query, fresh, 3, 3) == {0: [1.0, 2.0, 3.0], 1: [4.0, 5.0, 6.0]}
        finally:
            stage.release_query(query).wait()
            paths.destroy_path_list(plist)


@pytest.mark.skipif(
    not _NUMPY_HONORS_READONLY, reason="numpy < 2.1 ignores the DLPack read-only flag"
)
def test_map_group_dlpack_readonly_view_rejects_write(stage):
    # A read-only DLPack export must be refused by the consumer with a clean
    # ValueError. Regression guard: the DLPack deleter used to be a Python ctypes
    # callback, so releasing the tensor while that ValueError was propagating
    # aborted the callback at its first C call. The pending exception was reported
    # unraisable and cleared, the caller saw "SystemError: error return without
    # exception set" instead, and the callback's cleanup never ran.
    with PathDictionary(stage) as paths:
        order = ["/World/ReadOnly/A", "/World/ReadOnly/B"]
        plist = paths.create_path_list_from_strings(order)
        query = stage.query_from_path_list(plist)
        try:
            attr = paths.intern_token("map-readonly")
            stage.write_attribute(
                query, attr, ordinal=1, tensors=np.array([1.0, 2.0], np.float32), is_array=False
            ).wait()
            stage.advance_write_floor(ordinal=1).wait()

            with stage.map_attribute(query, attr, ordinal=2) as session:
                session.wait()
                for group in session.groups():
                    # [snippet:map-dlpack-readonly]
                    # A map group is writable, so dlpack() exports it writable by
                    # default and an in-place fill lands in the mapped buffer.
                    np.from_dlpack(group.dlpack(0, readonly=False))[:] = 5.0

                    # Pass readonly=True to hand out a view the recipient must not
                    # write through — numpy honors the DLPack read-only flag and
                    # refuses the assignment rather than corrupting the buffer.
                    with pytest.raises(ValueError):
                        np.from_dlpack(group.dlpack(0, readonly=True))[:] = 1.0
                    # [/snippet:map-dlpack-readonly]
                    session.unmap_group(group).wait()
            stage.advance_write_floor(ordinal=2).wait()

            # The refused write changed nothing: the committed column still holds
            # what the writable fill put there.
            assert _rows_by_prim(stage, query, attr, 2, 1) == {0: [5.0], 1: [5.0]}
        finally:
            stage.release_query(query).wait()
            paths.destroy_path_list(plist)

File: tests/c/test_map_attribute.cpp
// SPDX-FileCopyrightText: Copyright (c) 2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
// SPDX-License-Identifier: LicenseRef-NvidiaProprietary
//
// Public ovstage map/unmap test: map_attribute reserves writable storage-side
// buffers you fill directly (the zero-copy programming model), committed per group
// via unmap_group or all at once by the final unmap. Covers an existing column and
// a freshly-created one. CPU-only. The write-flavors example is the workflow tour.

#include <ovstage/ovstage.h>
#include <ovx/path_dictionary/path_dictionary.h>
#include <ovx/path_dictionary/path_dictionary_utils.h>
#include <dlpack/dlpack.h>

#include <cstdint>
#include <cstring>

#include <gtest/gtest.h>

namespace
{

ovx_string_t str(const char* s)
{
    return ovx_string_t{ s, std::strlen(s) };
}

bool waitOp(ovstage_instance_t* stage, ovstage_enqueue_result_t enq, const char* what)
{
    if (enq.status != OVSTAGE_OK)
    {
        ADD_FAILURE() << what << " enqueue rejected (code " << enq.status << ")";
        return false;
    }
    ovstage_op_wait_result_t wait{};
    const ovstage_api_status_t werr = ovstage_wait_op(stage, enq.op_index, OVSTAGE_TIMEOUT_INFINITE, &wait);
    for (size_t i = 0; i < wait.error_op_id_count; ++i)
    {
        const ovx_string_t e = ovstage_get_last_op_error(stage, wait.error_op_ids[i]);
        ADD_FAILURE() << what << " op failed: " << std::string(e.ptr ? e.ptr : "", e.ptr ? e.length : 0);
    }
    ovstage_release_op(stage, enq.op_index);
    if (werr != OVSTAGE_OK)
        ADD_FAILURE() << what << " wait failed (code " << werr << ")";
    return werr == OVSTAGE_OK && wait.error_op_id_count == 0;
}

class MapTest : public ::testing::Test
{
protected:
    void SetUp() override
    {
        ovstage_instance_desc_t desc{};
        desc.name = "test.ovstage.map";
        ASSERT_EQ(ovstage_create_instance(&desc, &stage_), OVSTAGE_OK);
        dict_ = ovstage_get_path_dictionary(stage_);
        ASSERT_NE(dict_, nullptr);
        const ovx_string_t paths[] = { str("/World/Mapped/A"), str("/World/Mapped/B") };
        ASSERT_EQ(path_dictionary_create_path_list_from_strings(dict_, paths, 2, &pathList_).status, OVX_API_SUCCESS);
        ASSERT_EQ(ovstage_query_from_path_list(stage_, pathList_, &query_), OVSTAGE_OK);
    }

    void TearDown() override
    {
        if (stage_ && query_ != OVSTAGE_INVALID_QUERY_HANDLE)
            waitOp(stage_, ovstage_release_query(stage_, query_), "release_query");
        if (dict_ && pathList_ != OVX_INVALID_PRIMPATH_LIST)
            path_dictionary_release_path_list_reference(dict_, pathList_);
        if (stage_)
            ovstage_destroy_instance(stage_);
    }

    void seal(ovstage_ordinal_t ordinal)
    {
        ovstage_write_floor_desc_t floor{};
        floor.ordinal = ordinal;
        floor.scope = OVSTAGE_SCOPE_ALL;
        ASSERT_TRUE(waitOp(stage_, ovstage_advance_write_floor(stage_, &floor), "advance_write_floor"));
    }

    // Read a fixed-lane column into out[list-index * lanes ...] in query order.
    void read(ovx_token_t attr, ovstage_ordinal_t endOrdinal, float* out, int lanes)
    {
        ovstage_ordinal_range_t range{};
        range.end_ordinal = endOrdinal;
        ovstage_read_handle_t r = OVSTAGE_INVALID_READ_HANDLE;
        ASSERT_TRUE(waitOp(stage_, ovstage_read_attributes(stage_, query_, &attr, 1, range, &r), "read"));
        for (;;)
        {
            ovstage_read_group_t group{};
            const ovstage_api_status_t fetched = ovstage_fetch_read_next(stage_, r, OVSTAGE_TIMEOUT_INFINITE, &group);
            if (fetched == OVSTAGE_ERROR_END_OF_ITERATION)
                break;
            ASSERT_EQ(fetched, OVSTAGE_OK);
            if (!group.is_delete && group.data.tensor_count == 1 && group.data.tensors[0].data)
            {
                const float* rows = static_cast<const float*>(group.data.tensors[0].data);
                for (uint32_t local = 0; local < group.prims.count; ++local)
                {
                    const size_t listIndex = group.prims.index_map ? group.prims.index_map[local]
                                                                   : group.prims.offset + local;
                    const uint32_t row = group.data.index_map ? group.data.index_map[local] : local;
                    for (int l = 0; l < lanes; ++l)
                        out[listIndex * lanes + l] = rows[row * lanes + l];
                }
            }
            ovstage_release_group(stage_, &group);
        }
        waitOp(stage_, ovstage_release_read(stage_, r), "release_read");
    }

    ovstage_instance_t* stage_ = nullptr;
    path_dictionary_instance_t* dict_ = nullptr;
    ovx_primpath_list_t pathList_ = OVX_INVALID_PRIMPATH_LIST;
    ovstage_query_handle_t query_ = OVSTAGE_INVALID_QUERY_HANDLE;
};

TEST_F(MapTest, MapExistingAndFreshColumn)
{
    const ovx_string_t names[] = { str("map-existing"), str("map-fresh") };
    ovx_token_t tokens[2] = { OVX_INVALID_TOKEN, OVX_INVALID_TOKEN };
    ASSERT_EQ(path_dictionary_create_tokens_from_strings(dict_, names, 2, tokens).status, OVX_API_SUCCESS);
    const ovx_token_t existing = tokens[0];
    const ovx_token_t fresh = tokens[1];

    float seed[] = { 1.0f, 2.0f };
    int64_t shape[] = { 2 };
    int64_t strides[] = { 1 };
    DLTensor seedTensor{ seed, { kDLCPU, 0 }, 1, { kDLFloat, 32, 1 }, shape, strides, 0 };
    ovstage_write_data_t seedWrite{};
    seedWrite.tensors = &seedTensor;
    seedWrite.tensor_count = 1;
    seedWrite.is_array = false;
    ASSERT_TRUE(waitOp(stage_,
        ovstage_write_attribute(stage_, query_, { existing, {} }, 1, seedWrite, OVSTAGE_PRIM_MODE_UPSERT), "seed"));
    seal(1);

    // [snippet:map-unmap-cpu-c]
    // map_attribute reserves storage-side buffers you fill directly. The mapped
    // buffer is a WRITE-ONLY staging buffer for the session's ordinal
    // (uninitialized, not a view of current values), so fill every element.
    // Commit per group via unmap_group (streaming) or all at once with the final
    // unmap_attribute, which also releases the session. CPU unmaps pass a zero
    // ovstage_cuda_sync_t (no GPU work to wait on).
    ovstage_map_desc_t existingDesc{};
    existingDesc.attribute = { existing, {} };
    existingDesc.prim_mode = OVSTAGE_PRIM_MODE_UPSERT;
    ovstage_map_handle_t map = OVSTAGE_INVALID_MAP_HANDLE;
    ASSERT_TRUE(waitOp(stage_,
        ovstage_map_attribute(stage_, query_, &existingDesc, 2, nullptr, 0, &map), "map existing"));
    for (;;)
    {
        ovstage_map_group_t mapGroup{};
        const ovstage_api_status_t fetched = ovstage_fetch_map_next(stage_, map, OVSTAGE_TIMEOUT_INFINITE, &mapGroup);
        if (fetched == OVSTAGE_ERROR_END_OF_ITERATION)
            break;
        ASSERT_EQ(fetched, OVSTAGE_OK);
        float* buffer = static_cast<float*>(mapGroup.data.tensors[0].data);
        for (uint32_t local = 0; local < mapGroup.prims.count; ++local)
        {
            const size_t listIndex = mapGroup.prims.index_map ? mapGroup.prims.index_map[local]
                                                              : mapGroup.prims.offset + local;
            const uint32_t row = mapGroup.data.index_map ? mapGroup.data.index_map[local] : local;
            buffer[row] = (listIndex == 0) ? 10.0f : 20.0f;
        }
        ASSERT_TRUE(waitOp(stage_, ovstage_unmap_group(stage_, map, &mapGroup, ovstage_cuda_sync_t{}), "unmap_group"));
    }
    ASSERT_TRUE(waitOp(stage_, ovstage_unmap_attribute(stage_, map, ovstage_cuda_sync_t{}), "unmap_attribute"));
    seal(2);

    // A NEW column: the descriptor dtype (3-lane float32) defines it. No
    // unmap_group this time — the final unmap commits everything.
    ovstage_map_desc_t freshDesc{};
    freshDesc.attribute = { fresh, {} };
    freshDesc.dtype = DLDataType{ kDLFloat, 32, 3 };
    freshDesc.prim_mode = OVSTAGE_PRIM_MODE_UPSERT;
    ASSERT_TRUE(waitOp(stage_,
        ovstage_map_attribute(stage_, query_, &freshDesc, 3, nullptr, 0, &map), "map fresh"));
    for (;;)
    {
        ovstage_map_group_t mapGroup{};
        const ovstage_api_status_t fetched = ovstage_fetch_map_next(stage_, map, OVSTAGE_TIMEOUT_INFINITE, &mapGroup);
        if (fetched == OVSTAGE_ERROR_END_OF_ITERATION)
            break;
        ASSERT_EQ(fetched, OVSTAGE_OK);
        float* buffer = static_cast<float*>(mapGroup.data.tensors[0].data);
        for (uint32_t local = 0; local < mapGroup.prims.count; ++local)
        {
            const size_t listIndex = mapGroup.prims.index_map ? mapGroup.prims.index_map[local]
                                                              : mapGroup.prims.offset + local;
            const uint32_t row = mapGroup.data.index_map ? mapGroup.data.index_map[local] : local;
            const float base = (listIndex == 0) ? 1.0f : 4.0f;
            buffer[row * 3 + 0] = base;
            buffer[row * 3 + 1] = base + 1.0f;
            buffer[row * 3 + 2] = base + 2.0f;
        }
    }
    ASSERT_TRUE(waitOp(stage_, ovstage_unmap_attribute(stage_, map, ovstage_cuda_sync_t{}), "unmap_attribute"));
    seal(3);
    // [/snippet:map-unmap-cpu-c]

    float existingValues[2] = {};
    read(existing, 2, existingValues, 1);
    EXPECT_FLOAT_EQ(existingValues[0], 10.0f);
    EXPECT_FLOAT_EQ(existingValues[1], 20.0f);

    float freshValues[6] = {};
    read(fresh, 3, freshValues, 3);
    const float expected[6] = { 1, 2, 3, 4, 5, 6 };
    for (int i = 0; i < 6; ++i)
        EXPECT_FLOAT_EQ(freshValues[i], expected[i]);
}

} // namespace
